gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageImpl.java

```diff
@@ -2248,6 +2275,67 @@ private boolean canIgnoreExceptionForEmptyObject(
     return false;
   }
 
+  @Override
+  public boolean isHnBucket(URI src) throws IOException {
+    String bucketName = src.getAuthority();
+    Boolean isEnabled = cache.getIfPresent(bucketName);
+    if (isEnabled != null) {
+      return isEnabled;
+    }
+
+    String prefix = src.getPath().substring(1);
+
+    StorageControlClient storageControlClient = lazyGetStorageControlClient();
+    GetStorageLayoutRequest request =
+        GetStorageLayoutRequest.newBuilder()
+            .setPrefix(prefix)
+            .setName(StorageLayoutName.format("_", bucketName))
+            .build();
+
+    try (ITraceOperation to = TraceOperation.addToExistingTrace("getStorageLayout.Hn")) {
+      boolean result =
+          storageControlClient.getStorageLayout(request).getHierarchicalNamespace().getEnabled();
+
+      logger.atFine().log("Checking if %s is HN enabled returned %s", src, result);
+
+      cache.put(bucketName, result);
+
+      return result;
+    }
+  }
+
+  private StorageControlClient lazyGetStorageControlClient() throws IOException {
+    if (this.storageControlClient == null) {
+      this.storageControlClient =
+          StorageControlClient.create(
+              StorageControlSettings.newBuilder()
+                  .setCredentialsProvider(
+                      FixedCredentialsProvider.create(new CredentialAdapter(this.credential)))
+                  .build());
+    }
+
+    return this.storageControlClient;
+  }
+
+  @Override
+  public void renameHnFolder(URI src, URI dst) throws IOException {
+    String bucketName = src.getAuthority();
+    String srcFolder = FolderName.of("_", bucketName, src.getPath().substring(1)).toString();
+    RenameFolderRequest request =
+        RenameFolderRequest.newBuilder()
+            .setDestinationFolderId(dst.getPath().substring(1))
+            .setName(srcFolder)
+            .build();
+
+    try (ITraceOperation to = TraceOperation.addToExistingTrace("renameHnFolder")) {
+      logger.atFine().log("Renaming HN folder (%s -> %s)", src, dst);
+      this.storageControlClient.renameFolderCallable().call(request).getDone();
```

Comment:
Recommend using:

```
Folder renamedFolder = storageControlClient.renameFolderOperationCallable().call(request);
```

Yao had raised this question with Java Cloud team a while back and this option will have the GAPIC poll until the operation is completed. IIUC this code will return the value of done but not wait in the case LRO takes longer than servicing the request to complete the rename.

yaqs/8226591078996770816

Reply:
From looking at the code, looks like this also block and waits for the future to finish. It is not clear to me what the return value of getDone signify.

Reply:
Pinged you offline; the take away here was provided by the GAPIC java team.

Reply:
Changed to renameFolderOperationCallable()